### DISCUSSION TITLE
convert from conll2003 to ner-json loses tags

### DISCUSSION THREAD

**Discussion by LadyHangaku** [explosion/spaCy#8738] [Help: Other Questions] [answered]
## How to reproduce the behaviour
command: `spacy convert filename.conll . --converter ner`

File must be a in a CoNLL-2003 format.

## Bug
When converting a file from CoNLL-2003 to an ner json file, a lot of the tags get lost in the process. Here are some snippets of the conll and ner files. In the first example, the tag gets lost, in the second one, with the same tags, the ner file maintained them:

CoNLL-2003:
```
Name _ O I-PER		
Last Name _ O I-PER
```

ner:
```
  "orth":"Name",
  "tag":"_",
  "ner":"O"
},
{
  "orth":"Last Name",
  "tag":"_",
  "ner":"O"
```
------------------------------------------

CoNLL-2003:
```
Another Name _ O I-PER
Another Last Name _ O I-PER
```

ner: 
```
  "orth":"Another Name ",
  "tag":"_",
  "ner":"B-PER"
},
{
  "orth":"Another Last Name",
  "tag":"_",
  "ner":"L-PER"
```

## My Environment
* **spaCy version:** 2.3.7
* **Platform:** Linux-4.15.0-144-generic-x86_64-with-debian-buster-sid
* **Python version:** 3.7.6

**Comment by adrianeboyd** [comment, marked as answer]
Hi, the problem is that this format doesn't support spaces in the tokens. The converter expects to be able to split the columns by whitespace and that the IOB annotation is in the same column in all lines after splitting, typically either always in the 2nd column or always in the 4th column depending on the dataset.

**Reply by LadyHangaku** [reply]
Thank you for your fast response!
Let me refactor the examples, because there are typically no spaces in the the tokens for one tag, I just wrote the tokens a little mindlessly, because I didn't want to use actual data:

CoNLL-2003:

```
Name _ O I-PER		
LastName _ O I-PER
```


ner:

  ```
"orth":"Name",
  "tag":"_",
  "ner":"O"
},
{
  "orth":"LastName",
  "tag":"_",
  "ner":"O"
```
------------------------------------------
CoNLL-2003:

```
AnotherName _ O I-PER
AnotherLastName _ O I-PER
```


ner:

 ```
 "orth":"AnotherName ",
  "tag":"_",
  "ner":"B-PER"
},
{
  "orth":"AnotherLastName",
  "tag":"_",
  "ner":"L-PER"
```

This is how it looks like more accurately. Every token is just one word without any spaces and if I understood you correctly, data like that should be no problem. But I still have the problem that some tags are maintained in the ner files and some aren't.
I also checked if it has something to do with text parts before or after the tagged tokens, but I wasn't able to find any pattern which could prove that.

**Reply by adrianeboyd** [reply]
Hi, if you can provide a full example plus the exact `spacy convert` command that shows this issue I can look into it, but I can't replicate it from these examples.

**Reply by LadyHangaku** [reply]
I can try, but I use confidential data, therefore I need to "neutralize" them first. This might take a while.
The command looks exactly like this, but with the specific filename:
`spacy convert filename.conll . --converter ner`
Thank you so much for your support!

**Reply by adrianeboyd** [reply]
As long you're not replacing whitespace, you can replace all the tokens with "x". It shouldn't matter for the conversion. It actually might be a good test, because if replacing them with "x" does solve the problem, then it's possibly something in the token encoding?

**Reply by LadyHangaku** [reply]
So I checked if the conversion to just x's as tokens makes a difference and it does.
But I still can't make out what the problem is in the encoding. The exact same name and last name sometimes keeps its tag, sometimes it doesn't. Here is an example from the data file. Maybe it has something to do with the tokens before/after the tagged ones?


CoNLL-2003:

```
• _ O O
Name _ O I-PER
LastName _ O I-PER
, _ O O
```

ner:

```
{
  "orth":"\u2022",
  "tag":"_",
  "ner":"O"
},
{
  "orth":"Name",
  "tag":"_",
  "ner":"B-PER"
},
{
  "orth":"LastName",
  "tag":"_",
  "ner":"L-PER"
},
{
  "orth":":",
  "tag":"_",
  "ner":"O"
},
```

--------------------------------------------

CoNLL-2003:

```
• _ O O
Name _ O I-PER
LastName _ O I-PER
: _ O O
```

ner:

```
{
"orth":"\u2022",
"tag":"_",
"ner":"O"
},
{
"orth":"Name",
"tag":"_",
"ner":"O"
},
{
"orth":"LastName",
"tag":"_",
"ner":"O"
},
{
"orth":":",
"tag":"_",
"ner":"O"
},
```

**Reply by adrianeboyd** [reply]
Hmm, I get the same results for both of those examples? What you probably need to check is that `line.split()` returns the same number of values for all the lines in your data.

**Reply by LadyHangaku** [reply]
Oh good idea! I just checked that and it seems like, beside the empty lines with only a "\n" in them, they all have len == 4... it's a mystery

**Reply by LadyHangaku** [reply]
Oh and the actual name used in the doc has no umlauts or some special chars.

**Reply by LadyHangaku** [reply]
Update:
I took a small part of the doc and made it an own file with round about 20 tokens. Then I used the -s flag and the conversion worked well. When I used -s on the whole original doc, again, it didn't work. So I suppose it has something to do with the document length?

**Reply by polm** [reply]
I really think that somewhere in your data there is a weird character that is messing up conversion, and it wasn't in the 20 token segment you took. Maybe try going over all 20 token segments or other natural chunks automatically and seeing if any of them cause an error/lose data to find the line at issue?

If you know you should never have `_` as a tag you should be able to automate a check for that.
